### DISCUSSION TITLE
Need help with Representation Context in new IfcAlignment example

### DISCUSSION THREAD

**Discussion by RickBrice** [IfcOpenShell/IfcOpenShell#4084] [Q&A]
I recently added an IfcAlignment project to the IFCOS Examples. I don't think I'm modeling the representation contexts correctly.

@aothms or others please review and advise.

**Comment by aothms** [comment]
I would copy the contexts from:

### Annex E.2.1 - Fixed Reference Swept Area Solid

https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/annex_e/alignment-geometries-and-linear-positioning/fixed-reference-swept-area-solid.html

```
#13   = IFCGEOMETRICREPRESENTATIONCONTEXT($, 'Model', 3, 1.E-4, #14, #16);
#17   = IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Body', 'Model', *, *, *, *, #13, $, .MODEL_VIEW., $);
#105  = IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Axis', 'Model', *, *, *, *, #13, $, .MODEL_VIEW., $);
#1001 = IFCGEOMETRICREPRESENTATIONSUBCONTEXT('FootPrint', 'Model', *, *, *, *, #13, $, .MODEL_VIEW., $);
```

So concretely for the example I would suggest:

- create two sub contexts named Axis and Footprint, now you only have Axis.
- `GRAPH_VIEW` vs `MODEL_VIEW` vs `PLAN_VIEW`. I to be honest don't know. The example just uses MODEL_VIEW. I'm not aware of any concrete implications of trying to better differentiate this. None of these match perfectly as these were cooked up before infra was added (we did have grids though, but nobody ever specified whether the representation of a grid is Axis/Footprint/Body and whether it deserves a separate non-model-view context. You could bring this up at the impl forum, but it's really a bit of a consequence-free semantics.
- I would use `"Model"` instead of `"3D"` for ContextType.

**Comment by RickBrice** [comment]
Thanks @aothms. Following your advice and trying to implement as I understand the CTs for Alignment Geometry, I've updated the example.

I'm trying to generate geometry that can be viewed in Blender from an OBJ file. Using the commands for IfcConvert from our other [discussion](https://github.com/IfcOpenShell/IfcOpenShell/pull/4023#issuecomment-1822296204), I was getting error that 3D wasn't correct (don't recall exact error message).

With the new model, I get a bit more progress but a crash in IfcConvert.

Here is the command line I'm using along with the console output (using PowerShell console)
~~~
PS F:\IfcOpenshell\_build-vs2022-x64\Debug> .\IfcConvert.exe F:\IfcOpenShell\_build-vs2022-x64\examples\FHWA_Bridge_Geometry_Alignment_Example.ifc -yvv FHWA_Bridge_Geometry_Alignment_Example.obj --plan --include attribute GlobalId 2RLFlDMxj2a9KshyXnMaip
IfcOpenShell IfcConvert 0.8.0 (OCC 7.7.1)
[Notice] [2023-12-07 09:36:15] exclude entities IfcOpeningElement IfcSpace
[Notice] [2023-12-07 09:36:15] include GlobalId values "2RLFlDMxj2a9KshyXnMaip"
Scanning file...
Done scanning file
Parsing input file took 0 seconds
~~~

The debugger emits
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/ffefa128-aaa4-43aa-9686-2b9302d31eba)

The call stack is
~~~
IfcConvert.exe!std::_Vector_const_iterator<std::_Vector_val<std::_Simple_types<std::shared_ptr<ifcopenshell::geometry::taxonomy::geom_item>>>>::operator--() Line 94
	at C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\vector(94)
IfcConvert.exe!std::_Vector_iterator<std::_Vector_val<std::_Simple_types<std::shared_ptr<ifcopenshell::geometry::taxonomy::geom_item>>>>::operator--() Line 296
	at C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\vector(296)
IfcConvert.exe!ifcopenshell::geometry::filter_in_place<`ifcopenshell::geometry::mapping_Ifc4x3_add2::map_impl'::`2'::<lambda_1>>(std::shared_ptr<ifcopenshell::geometry::taxonomy::collection> collection, ifcopenshell::geometry::mapping_Ifc4x3_add2::map_impl::__l2::<lambda_1> fn) Line 1267
	at F:\IfcOpenshell\src\ifcgeom\taxonomy.h(1267)
IfcConvert.exe!ifcopenshell::geometry::mapping_Ifc4x3_add2::map_impl(const Ifc4x3_add2::IfcRepresentation * inst) Line 50
	at F:\IfcOpenshell\src\ifcgeom\mapping\IfcRepresentation.cpp(50)
IfcConvert.exe!ifcopenshell::geometry::mapping_Ifc4x3_add2::map(const IfcUtil::IfcBaseInterface * inst) Line 31
	at F:\IfcOpenshell\src\ifcgeom\mapping\mapping.i(31)
IfcConvert.exe!IfcGeom::Iterator::initialize() Line 177
	at F:\IfcOpenshell\src\ifcgeom\Iterator.h(177)
IfcConvert.exe!wmain(int argc, wchar_t * * argv) Line 979
	at F:\IfcOpenshell\src\ifcconvert\IfcConvert.cpp(979)
~~~

Is the problem my model or something with IfcConvert?
[FHWA_Bridge_Geometry_Alignment_Example.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/13603295/FHWA_Bridge_Geometry_Alignment_Example.ifc.txt)

**Comment by RickBrice** [comment]
I ran that example file, which I think conforms to the requirements of the specification and concept templates, through the validation service. The schema syntax check fails on IfcShapeModel.WR11 because there is more than one product representation.

https://validate.buildingsmart.org/report_syntax_schema/WGktvLuRPdnEfqOzhzDsvumtpWaQkqfJ

Also see https://github.com/buildingSMART/IFC4.3.x-development/issues/742.

Is my model all messed up? What am I not understanding?

**Comment by RickBrice** [comment]
I kept tweaking the example until I got it to pass the bSI validation service. Here is the updated file. It still crashes IfcConvert.

It looks to me like there is a disconnection between some of the validation rules and the concept template documentation. At one point, I got the schema rules to pass, but then violation of the GEM003 rule when trying to satisfy the requirements of both Alignment Geometry - Horizontal and Alignment Geometry - Horizontal and Vertical. This may be of interest to you since you are working on the validation service (I think). You probably have access to the files I used and the validation outcome. If not, I can get them for you.

Perhaps only one concept template can be applied at a time? As far as I can see, the documentation is silent on this.
 
[FHWA_Bridge_Geometry_Alignment_Example.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/13606582/FHWA_Bridge_Geometry_Alignment_Example.ifc.txt)

**Comment by aothms** [comment]
In my case I get an infinite recursion due to this line:

```
#20=IFCLOCALPLACEMENT(#20,#19);
```

With just that changed to

```
#20=IFCLOCALPLACEMENT($,#19);
```

.. I get some output, so that's really exciting and promising!

![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/ac36aca6-c1b9-4a8d-bd4f-1e7a3f3ea7da)

I guess this come from the following:

```C++
auto local_placement = site->ObjectPlacement();
if (!local_placement) {
  local_placement = file.addLocalPlacement();
}

...

auto alignment = new Schema::IfcAlignment(IfcParse::IfcGlobalId(), nullptr, std::string("Example Alignment"), boost::none, boost::none, local_placement, alignment_product, boost::none);

...

file.relatePlacements(site, alignment);
```

You directly assign the site placement to the alignment and then ask for them to be related, which apparently creates this.

We should probably do a check in IfcHierarchyHelper::relatePlacements that `parent->ObjectPlacement() != local_place`. But at that time when it was written for IFC2X3 reusing placements wasn't allowed yet as the inverse was [0..1] cardinality.

**Reply by RickBrice** [reply]
@aothms I found that issue late yesterday. The cause is exactly as you suggested. 

I'll look into adding a catch for recursive placements. 

The graphic is what I expected. I'll create it in my own and will try to figure out was is going on. 

Thanks for your help.

**Comment by RickBrice** [comment]
@aothms The IFC I was generating for the IfcAlignment example was somehow incorrect. When I open the file in other viewers like RDFs IfcViewer or ODAs Open IFC Viewer, the alignment geometry was a complete mess.

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/d12c99aa-801b-477c-85fa-333d2bb79095)

After days of trial and error, I finally get something reasonable.

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/a4f68039-1058-41d1-bf09-731c270b3c89)

From there, I had to update the internals of IfcCurveSegment to get the geometry to come out right from IFCOS (I was previously computing geometry based on my flawed understanding of IFC).

So, I'm back to trying to use IfcConvert to get something I can open in Blender.

I get the same errors as before (error with vector iterator) with this command and attached file. No geometry is output in an OBJ file. Actually, there is not an OBJ file.

~~~
 F:\IfcOpenshell\_build-vs2022-x64\Debug> .\IfcConvert.exe F:\IfcOpenShell\_build-vs2022-x64\examples\FHWA_Bridge_Geometry_Alignment_Example.ifc -yvv FHWA_Bridge_Geometry_Alignment_Example.obj --plan --include attribute GlobalId 0zjB8EfuTCJex3g3E4doQC
~~~
[FHWA_Bridge_Geometry_Alignment_Example.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/13675943/FHWA_Bridge_Geometry_Alignment_Example.ifc.txt)

I get the same error for other files I generate and from the linear-placement-of-signal.ifc example from the IFC documentation.

I'm using IfcConvert from the debug build on the v0.8.0 branch.

Any suggestions?

My goal is to be able to export IFC for alignment and bridge model from my software (PGSuper) and be able to open it in BlenderBIM. At first for just viewing and later for editing and manipulating. Am I even close to being on the right track?

**Reply by aothms** [reply]
Wow that's a setback, but impressive work.

Good news is that here on my side IfcConvert works now also on the alignment output, so no issues there.

![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/e355abe2-1999-452b-9d2a-cba2bbe152b2)

---

Regarding the issue, sorry, I think I finally understand now, it took me a while to realize that in an msvc debug build (so not just symbols what I normally use with CMake's RelWithDebInfo, but the actual Debug with `_ITERATOR_DEBUG_LEVEL = 2`) there is much more extensive iterator validation. So our logic there is indeed not valid, even though we'd never dereference the it, even the act of decrementing before begin() is not allowed. It's running late here and changing taxonomy.h causes a full rebuild so I wasn't able to test (truly sorry) but we probably have to rewrite to a more conventional remove-erase implementation:

```C++
		// @nb traverses nested collections
		template <typename Fn>
		taxonomy::collection::ptr filter_in_place(taxonomy::collection::ptr collection, Fn fn) {
			auto& c = collection->children;
			auto new_end = std::remove_if(c.begin(), c.end(), [](taxonomy::geom_item::ptr i) {
				return !apply_predicate_to_collection(i, fn);
			});
#ifdef TAXONOMY_USE_NAKED_PTR
			for (auto it = new_end; it != c.end(); ++it) {
				delete *it;
			}
#endif
			c.erase(new_end, c.end());
			return collection;
		}
```

**Comment by RickBrice** [comment]
I'm happy to see the geometry in Blender. That alone fortifies my motivation to keep going.

**Reply by aothms** [reply]
Me as well. I committed something for filter_in_place() hope it works now on Debug builds as well

**Comment by RickBrice** [comment]
I apologize that my python is a little rusty. After I get a build, what do I need to copy into F:\IfcOpenShell\_deps-vs2022-x64-installed\Python34\Lib\site-packages in order to use IFCOS from python?

Or maybe a more fundamental question - if I had IFCOS v0.8 set up correctly for python, are the C++ bindings automatically generated so I can call the alignment geometry features from python?

**Reply by aothms** [reply]
No worries. If you run the `INSTALL` target you have in MSVC it will automatically package everything under that site-packages folder.

And yes, you should be able to do things like:

```python
import numpy
import ifcopenshell
import ifcopenshell.geom

f = ifcopenshell.open(...)
s = ifcopenshell.geom.settings(PLAN=True)
for inst in f.by_type('IfcAlignment'):
    shp = ifcopenshell.geom.create_shape(s, inst)
    vs = numpy.array(shp.geometry.verts).reshape((-1,3))
    es = numpy.array(shp.geometry.edges).reshape((-1,2))
    print(vs[es])
```

It's a little bit unfortunate that the default python interpreter that comes from the win build script is somewhat antique. That might give you some difficulties.

It's no too hard to configure a custom python interpreter through cmake (or just edit the CMakeCache.txt files and regenerate). But if you don't intend to rely on too many other modules then the python34 from the build script should also work.

**Comment by RickBrice** [comment]
@aothms Not sure if then changes you recently made are the culprit, but I'm now getting build errors for the msvc Debug configuration

~~~
2>------ Build started: Project: _ifcopenshell_wrapper, Configuration: Debug x64 ------
2>LINK : warning LNK4075: ignoring '/INCREMENTAL' due to '/FORCE' specification
2>libhdf5_cpp.lib(H5IdComponent.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in IfcPythonPYTHON_wrap.obj
2>libhdf5_cpp.lib(H5IdComponent.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in IfcPythonPYTHON_wrap.obj
2>libhdf5_cpp.lib(H5Group.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in IfcPythonPYTHON_wrap.obj
2>libhdf5_cpp.lib(H5Group.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in IfcPythonPYTHON_wrap.obj
2>libhdf5_cpp.lib(H5IdComponent.obj) : warning LNK4099: PDB '' was not found with 'libhdf5_cpp.lib(H5IdComponent.obj)' or at ''; linking object as if no debug info
2>libhdf5_cpp.lib(H5Group.obj) : warning LNK4099: PDB '' was not found with 'libhdf5_cpp.lib(H5Group.obj)' or at ''; linking object as if no debug info

... and others of the same two types
~~~

**Reply by aothms** [reply]
>     2>libhdf5_cpp.lib(H5IdComponent.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in IfcPythonPYTHON_wrap.obj

>     2>libhdf5_cpp.lib(H5IdComponent.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in IfcPythonPYTHON_wrap.obj

So apparently `libhdf5_cpp.lib` is a Release compiled archive, but `IfcPythonPYTHON_wrap.obj` is a Debug compiled archive. `_ITERATOR_DEBUG_LEVEL` 0 vs 2 and `MD_DynamicRelease` vs `MDd_DynamicDebug`.

Somehow cmake decided to link the release hdf5 libs into your debug python wrapper.

I recall you mentioned earlier you made some changes somewhere or had issues searching for the HDF5 libraries. This must be related to what we're seeing here.

**Comment by RickBrice** [comment]
Also having trouble getting IFCOS working with python. I build the RelWithDebugInfo configuration and ran the install script.

Here is what I get

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/c54a1877-47c8-4a26-a85c-e0402b9eb8fb)


Any ideas?

**Reply by aothms** [reply]
Oh oops sorry yeah it looks like Python 3.4 is really not in anyway supported anymore because of the usage of f-strings in the codebase. I propose:

(a) either installing a recent Python version from the web and then

```
set PYTHONHOME=C:\Python39
set PY_VER_MAJOR_MINOR=39
```

Call `run-cmake.bat` with the appropriate args

(b) or edit

`IfcOpenShell\_build-vs2019-x64\CMakeCache.txt` and update all `PYTHON_...` variables accordingly. Then make a meaningless change in IfcOpenShell\CMakeLists.txt` so that the cmake script in MSVC is reran (you can right click on it and compile, but it has to have been changed). Then reload the MSVC solution and the paths will have updated.

**Reply by RickBrice** [reply]
OK - thanks. I'll upgrade my Python and try again.
Probably wont get much done over the next couple of weeks so I thought I'd just play around with python.

**Comment by RickBrice** [comment]
@aothms sorry to bother you - I'm close, but still can't get a python version of the v0.8.0 running.

When I run `run-cmake.bat` I get the following error:

~~~-- Looking for Python header files in: F:/Python/Python312/include
-- Looking for Python library file in: F:/Python/Python312/libs/python312.lib
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'distutils'
CMake Warning at F:/IfcOpenShell/src/ifcwrap/CMakeLists.txt:90 (MESSAGE):
  Unable to locate Python site-package directory, unable to install the
  Python wrapper
~~~

I tried to `pip install distutils` but get
~~~
F:\Python\Python312> pip install distutils
ERROR: Could not find a version that satisfies the requirement distutils (from versions: none)
ERROR: No matching distribution found for distutils
~~~

Any ideas?

**Reply by aothms** [reply]
You're very close. On v0.7 we have a fix for this (supposedly, I didn't test) but it hasn't been merged into v0.8 yet https://github.com/IfcOpenShell/IfcOpenShell/commit/eb91f5179b4c446a0541d37ec094483e537dc811

**Comment by RickBrice** [comment]
I cherry picked that commit to the v0.8 branch - it work great.

Running the C++ alignment code from Python

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/07a876fc-725b-45c8-b583-31f5bb25ff16)

Visualizing the geometry in Blender
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/c9e589df-1d89-45ab-a3c6-8b407262677a)

I know there is still a lot of little things to address, but maybe after the first of the year we can have an online meetup and discuss what the next steps might look like. I'm thinking how to get BlenderBIM using the v0.8.0 library and opening/editing alignment models - maybe it's too early for that.
